sdk/go/provision_test.go

```diff
@@ -0,0 +1,100 @@
+package dagger
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"dagger.io/dagger/internal/engineconn/dockerprovision"
+	"github.com/adrg/xdg"
+	"github.com/moby/buildkit/identity"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+func TestImageProvision(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	daggerHost, ok := os.LookupEnv("DAGGER_HOST")
+	if ok {
+		if !strings.HasPrefix(daggerHost, dockerprovision.DockerImageConnName+"://") {
+			t.Skip("DAGGER_HOST is not set to docker-image://")
+		}
+	}
+
+	tmpdir := t.TempDir()
+	os.Setenv("XDG_CACHE_HOME", tmpdir)
+	defer os.Unsetenv("XDG_CACHE_HOME")
+	xdg.Reload()
+	cacheDir := filepath.Join(tmpdir, "dagger")
+
+	// create some garbage for the image provisioner to collect
+	err := os.MkdirAll(cacheDir, 0700)
+	require.NoError(t, err)
+	f, err := os.Create(filepath.Join(cacheDir, "dagger-engine-session-gcme"))
+	require.NoError(t, err)
+	f.Close()
+
+	tmpContainerName := "dagger-engine-gcme-" + identity.NewID()
+	if output, err := exec.CommandContext(ctx,
+		"docker", "run",
+		"--rm",
+		"--detach",
+		"--name", tmpContainerName,
+		"busybox",
+		"sleep", "120",
+	).CombinedOutput(); err != nil {
+		t.Fatalf("failed to create container: %s", output)
+	}
+
+	parallelism := 30
+	start := make(chan struct{})
+	var eg errgroup.Group
+	for i := 0; i < parallelism; i++ {
+		eg.Go(func() error {
+			<-start
+			c, err := Connect(ctx)
+			if err != nil {
+				return err
+			}
+			defer c.Close()
+			// do a trivial query to ensure the engine is actually there
+			_, err = c.Container().From("alpine:3.16").ID(ctx)
```

Comment:
@aluzzardi this is a case where the catch-22 we talked about before could happen: if we ever change this API then it won't work since this test case by default runs against a release engine.

I considered setting up what we talked about in terms of actually pushing the dev dagger engine to a local registry, but I'm not sure if it's worth it yet since we have open PRs changing how we build+test with that anyways, so I left it out of this PR for now. Let me know if you think it's worth an immediate fix though.